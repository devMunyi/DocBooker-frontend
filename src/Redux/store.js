import { configureStore, applyMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import doctorReducer from './reducers/doctors';
import reservationReducer from './reducers/reservations';
import selectedDoctorReducer from './reducers/selectedDoctor';
import allDoctors from './reducers/allDoctors';
import user from './reducers/user';


export default configureStore(
  {
    reducer: {
      doctor: doctorReducer,
      reservations: reservationReducer,
      selectedDoctor: selectedDoctorReducer,
      doctors: allDoctors,
      user,
    },
  },
  applyMiddleware(logger),
);
