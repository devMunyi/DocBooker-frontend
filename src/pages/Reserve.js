import React, { useState, useEffect } from 'react';
import { Link, useParams, useNavigate } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import axios from 'axios';
import { DatePicker } from 'antd';
import { toast } from 'react-toastify';
import { SyncOutlined } from '@ant-design/icons';
import moment from 'moment';
import { fetchSingleDoctor } from '../Redux/actions/doctors';

function Reserve() {
  const [isProcessing, setIsProcessing] = useState(false);
  const [date, setDate] = useState('');

  const dispatch = useDispatch();
  const navigate = useNavigate();
  const { doctorId } = useParams(); // get doctor's id from url
  const userId = JSON.parse(localStorage.getItem('user')).id; // grab logged in user id

  // get single doctor info form redux store
  const { singleDoctor, isFetchingData } = useSelector(
    (state) => state?.doctor,
  );

  // fetch a single doctor
  useEffect(() => {
    dispatch(fetchSingleDoctor({ doctorId, userId }));
  }, [dispatch, doctorId, userId]);

  const handleDateChange = (value, dateString) => {
    console.log('Selected Time: ', value);
    setDate(dateString);
  };

  localStorage.setItem('doctor', JSON.stringify(singleDoctor));

  const handleBook = async () => {
    try {
      setIsProcessing(true);
      const response = await axios.post(
        `http://localhost:3000/api/users/${userId}/doctors/${doctorId}/reservations`,
        {
          reservation: {
            user_id: userId,
            doctor_id: doctorId,
            date,
            details: 'Some details',
          },
        },
        {
          headers: {
            'Content-Type': 'application/json',
          },
        },
      );
      const { data } = response;
      const { id } = data;
      if (response.status === 201) {
        toast.success('Reservation booked successfully!');
        setTimeout(() => {
          navigate(`/reservations/${doctorId}/${id}`);
        }, 2500);
      } else {
        throw new Error(data.message || 'Failed to book reservation');
      }
    } catch (error) {
      toast.error(error.message || 'Failed to book reservation');
    } finally {
      setIsProcessing(false);
    }
  };

  return (
    <div id="reserve-page" className="reservePage">
      <div className="container-fluid reservePage">
        {isFetchingData ? (
          <div
            className="d-flex justify-content-center align-items-center"
            style={{
              width: '100%',
              height: '100vh',
              backgroundColor: '#589d1a',
            }}
          >
            <SyncOutlined spin style={{ fontSize: '32px' }} />
          </div>
        ) : (
          <>
            <img
              src={`${singleDoctor?.picture}`}
              alt="Doctor"
              style={{ width: '100%', height: '100vh' }}
            />
            <div className="centered">
              <h1 className="doc-name pb-2">{`Dr ${singleDoctor?.name} - ${singleDoctor?.specialization}`}</h1>
              <p>
                Lorem, ipsum dolor sit amet consectetur adipisicing elit. Minus
                provident et expedita perferendis dolore perspiciatis odio.
                Dignissimos, in deleniti pariatur alias odit explicabo vitae
                libero provident est placeat consectetur hic?
              </p>

              <div className="d-flex gap-3 justify-content-center">
                <DatePicker
                  format="YYYY-MM-DD"
                  onChange={handleDateChange}
                  className="date-picker rounded-0"
                  disabledDate={(current) => current && current.valueOf() < moment().subtract(1, 'days')}
                />
              </div>

              <div className="d-flex gap-3 justify-content-center mt-4">
                <Link
                  to={`/doctors/${doctorId}`}
                  className="btn btn-outline-light pb-0"
                >
                  Go Back
                </Link>
                {isProcessing ? (
                  <button
                    className="btn btn-outline-light"
                    type="button"
                    disabled
                  >
                    <span
                      className="spinner-border spinner-border-sm"
                      role="status"
                      aria-hidden="true"
                    />
                    <span className="visually-hidden">Loading...</span>
                  </button>
                ) : (
                  <button
                    type="submit"
                    onClick={handleBook}
                    className="btn btn-outline-light"
                  >
                    Book Now
                  </button>
                )}
              </div>
            </div>
          </>
        )}
      </div>
    </div>
  );
}

export default Reserve;
