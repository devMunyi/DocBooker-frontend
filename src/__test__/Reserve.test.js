import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { MemoryRouter as Router } from 'react-router-dom';
import store from '../Redux/store';
import Reserve from '../pages/Reserve';

it('Reserve component renders correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Router>
          <Reserve />
        </Router>
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
