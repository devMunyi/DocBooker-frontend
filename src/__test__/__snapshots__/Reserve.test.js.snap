// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reserve component renders correctly 1`] = `
<div
  className="reservePage"
  id="reserve-page"
>
  <div
    className="container-fluid reservePage"
  >
    <img
      alt="Doctor"
      src="undefined"
      style={
        Object {
          "height": "100vh",
          "width": "100%",
        }
      }
    />
    <div
      className="centered"
    >
      <h1
        className="doc-name pb-2"
      >
        Dr undefined - undefined
      </h1>
      <p>
        Lorem, ipsum dolor sit amet consectetur adipisicing elit. Minus provident et expedita perferendis dolore perspiciatis odio. Dignissimos, in deleniti pariatur alias odit explicabo vitae libero provident est placeat consectetur hic?
      </p>
      <div
        className="d-flex gap-3 justify-content-center"
      >
        <div
          className="ant-picker css-dev-only-do-not-override-10ed4xt date-picker rounded-0"
          onClick={[Function]}
        >
          <div
            className="ant-picker-input"
          >
            <input
              autoComplete="off"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              placeholder="Select date"
              readOnly={true}
              size={12}
              title=""
              value=""
            />
            <span
              className="ant-picker-suffix"
            >
              <span
                aria-label="calendar"
                className="anticon anticon-calendar"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="calendar"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                  />
                </svg>
              </span>
            </span>
          </div>
        </div>
      </div>
      <div
        className="d-flex gap-3 justify-content-center mt-4"
      >
        <a
          className="btn btn-outline-light pb-0"
          href="/doctors/undefined"
          onClick={[Function]}
        >
          Go Back
        </a>
        <button
          className="btn btn-outline-light"
          onClick={[Function]}
          type="submit"
        >
          Book Now
        </button>
      </div>
    </div>
  </div>
</div>
`;
